### PR TITLE
Symbolic stoichiometric coefficients

### PR DESCRIPTION
**Pull request by isaacsas**
TODO:
- [x] ODEs and SDEs with tests.
- [x] Jumps with tests ~(needs ModelingToolkit JumpSystem updates)~ edit: Not needed apparently.
- [x] drop requirement for `use_combinatoric_ratelaws=false`.
- [x] benchmark to see if this significantly impacts system construction times.
- [ ] ~DSL support.~ edit: Will save for a future PR as this will be some work...
- [x] tests
- [x] check BCR with jumps for performance regressions on construction.

### CODE REVIEW

**Comment by isaacsas** [comment]
Testing BCR loading with this PR seems to not have any real impact on timings for loading and converting systems:
```julia
Current Catalyst:

prnbng = loadrxnetwork(BNGNetwork(),fname); 
522.819 ms (4224319 allocations: 189.59 MiB)

bosys = convert(ODESystem, rnbng)
2.448 s (26007243 allocations: 1.36 GiB)

boprob = ODEProblem(bosys, Float64[], (0.,tf), Float64[])
1.359 s (13747759 allocations: 662.72 MiB)

—————————————————
PR:
prnbng = loadrxnetwork(BNGNetwork(),fname)
509.176 ms (4224319 allocations: 189.59 MiB)

bosys = convert(ODESystem, rnbng)
2.456 s (26007243 allocations: 1.36 GiB)

boprob = ODEProblem(bosys, Float64[], (0.,tf), Float64[])
1.358 s (13747759 allocations: 662.72 MiB)
```
I forgot to interpolate into the `@btime` calls, but when I reran the PR code interpolating variables it had no real effect.

**Comment by isaacsas** [comment]
Need https://github.com/SciML/ModelingToolkit.jl/issues/1475 fixed for this to work with our normal rate laws since otherwise parameters get converted to float and factorials can't be evaluated.

**Comment by isaacsas** [comment]
Currently this should work for ODEs and SDEs. I may just update the DSL and then merge this. 

Everything for jumps should be in place here, but we need updates to JumpSystems in ModelingToolkit too which will have to be for another day.

**Comment by isaacsas** [comment]
BCR JumpSystem construction timings are also unaffected:
```
Jumps
Current Catalyst:
prn = loadrxnetwork(BNGNetwork(), fname)
504.861 ms (4215741 allocations: 189.36 MiB) 

jsys = convert(JumpSystem, rn)
53.115 ms (611173 allocations: 26.10 MiB)

dprob = DiscreteProblem(jsys, u0, (0.0,tequil), [])
6.167 ms (64467 allocations: 2.69 MiB)

jprob = JumpProblem(jsys, dprob, RSSACR(), save_positions=(false,false))
2.703 s (2701719 allocations: 255.82 MiB)

—————————————————
PR:
prn = loadrxnetwork(BNGNetwork(), fname)
501.691 ms (4215741 allocations: 189.36 MiB) 

jsys = convert(JumpSystem, rn)
49.913 ms (611173 allocations: 26.10 MiB)

dprob = DiscreteProblem(jsys, u0, (0.0,tequil), [])
5.995 ms (64467 allocations: 2.69 MiB)

jprob = JumpProblem(jsys, dprob, RSSACR(), save_positions=(false,false))
2.720 s (2701719 allocations: 255.82 MiB)
```

**Comment by isaacsas** [comment]
@ChrisRackauckas I think this is good to go, but if you could look it over when you have a chance to make sure I'm not doing something bad/dumb in my handling of symbolics that would be great.

File: src/reactionsystem.jl [hunk 1]
```diff
@@ -42,7 +42,7 @@ Notes:
 - The three-argument form assumes all reactant and product stoichiometric coefficients
   are one.
 """
-struct Reaction{S, T <: Number}
```
Comment (ChrisRackauckas) [inline src/reactionsystem.jl:45]: This change shouldn't be required, but that's fine.

**Review by ChrisRackauckas — APPROVED** [review]
Looks great.

**Pull request merged**